[1기-P 김명환] JPA 게시판 만들기

samkimuel opened the pull request:
## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

SpingBoot JPA 게시판 만들기

[README](https://github.com/samkimuel/SpringBoot-Board/blob/main/README.md)

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->

- 유저
    - 회원가입
    - 회원 정보 조회
    - 회원 정보 수정
    - 탈퇴

- 게시판
    - 게시글 페이징 조회(다건)
    - 게시글 상세 조회(단건)
    - 게시글 수정
    - 게시글 삭제

- REST DOCS API 문서화

- User <- Post 1:N 단방향 매핑

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

- Controller와 Request DTO, Response DTO 설계를 어떻게 하면 깔끔하게 만들 수 있을까요?

samkimuel committed:
Initial commit

samkimuel committed:
feat: Entity 설계, 구현

samkimuel committed:
chore: MySQL 의존성 설정, MySQL 연동 설정

samkimuel committed:
feat: Post 조회 API 개발

- 페이징 조회
- 단건 조회
- 예외 처리
- Request, Response DTO 구현

samkimuel committed:
refactor: @EnabledJpaAuditing 위치 수정

samkimuel committed:
refactor: Entity GeneratedValue strategy를 IDENTITY로 변경

samkimuel committed:
feat: 게시글 작성 API 개발, 유저 회원가입 API 개발

- UserNotFoundException 예외 처리 추가

samkimuel committed:
feat: 게시글 수정 API 개발

samkimuel committed:
feat: 게시글 삭제 API 개발

samkimuel committed:
feat: 유저 정보 조회, 유저 정보 수정, 유저 탈퇴 API 개발

- User 테이블 물리 삭제를 위해 FK 제약조건 제거

samkimuel committed:
refactor: PostConverter, UserConverter을 Bean으로 등록

- PostService, UserService에 Converter 의존성 주입

samkimuel committed:
refactor: Post, User의 Entity와 DTO 클래스의 생성자 수정

- Entity 클래스 생성자에 id 파라미터 추가
- DTO 생성자 추가

samkimuel committed:
fix: UserNotFoundException 예외 처리 코드 오류 수정

samkimuel committed:
refactor: 게시글 삭제 API return type을 DTO로 변경

samkimuel committed:
refactor: 유저 탈퇴 API return type을 DTO로 변경

samkimuel committed:
test: User Service, Controller 단위테스트 코드 작성

samkimuel committed:
test: Post Service, Controller 단위테스트 코드 작성

samkimuel committed:
docs: REST DOCS를 통한 API 문서화

samkimuel committed:
docs: README.md 수정

samkimuel commented on src/main/java/com/devcourse/springbootboard/post/controller/PostController.java:
```diff
@@ -0,0 +1,54 @@
+		this.postService = postService;
+	}
+
+	@GetMapping
+	public ResponseEntity<Page<PostResponse>> getAll(final Pageable pageable) {
+		return ResponseEntity.ok(postService.findPosts(pageable));
+	}
```
거기까지 생각을 못했는데 알려주셔서 감사합니다:) 구현해보겠습니다!

neilsonT commented on src/main/java/com/devcourse/springbootboard/user/controller/UserController.java:
```diff
@@ -0,0 +1,50 @@
+		return ResponseEntity.ok(userService.findUser(id));
+	}
+
+	@PutMapping("/{id}/profile")
+	public ResponseEntity<UserResponse> modifyUser(
+		@PathVariable final Long id,
+		@RequestBody final UserUpdateRequest userUpdateRequest
+	) {
+		return ResponseEntity.ok(userService.updateUser(id, userUpdateRequest));
+	}
```
/{id}/profile 에서 마지막에 profile을 사용하도록 정한 이유는 무엇인가용?

neilsonT commented on src/main/java/com/devcourse/springbootboard/user/domain/Hobby.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+
+@Embeddable
+@Getter
+@EqualsAndHashCode
+public class Hobby {
```
Hobby 클래스에 @EqualsAndHashCode를 사용한 이유는 무엇인가요?

neilsonT replied:
[롬복 주의사항](https://kwonnam.pe.kr/wiki/java/lombok/pitfall) 롬복 관련 좋은 글이 있더라구요~

neilsonT commented on src/main/java/com/devcourse/springbootboard/user/domain/User.java:
```diff
@@ -0,0 +1,49 @@
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
+	private Long id;
+
+	@Column(name = "name", nullable = false)
+	private String name;
+
+	@Column(name = "age", nullable = false)
+	private int age;
+
+	@Embedded
+	private Hobby hobby;
```
name과 age는 primitive 타입을 사용하고, hobby는 VO로 분리한 이유가 궁금합니다~

neilsonT commented on src/test/java/com/devcourse/springbootboard/post/controller/PostControllerTest.java:
```diff
@@ -0,0 +1,225 @@
+					parameterWithName("page").description("페이지 번호"),
+					parameterWithName("size").description("게시글 갯수")
+				),
+				responseFields(
+					fieldWithPath("content[].postId").type(JsonFieldType.NUMBER).description("게시글 아이디"),
+					fieldWithPath("content[].title").type(JsonFieldType.STRING).description("글 제목"),
+					fieldWithPath("content[].content").type(JsonFieldType.STRING).description("글 내용"),
+					fieldWithPath("content[].userId").type(JsonFieldType.NUMBER).description("작성자 아이디"),
+					fieldWithPath("pageable").type(JsonFieldType.STRING).description("pageable"),
+					fieldWithPath("totalPages").type(JsonFieldType.NUMBER).description("전체 페이지 수"),
+					fieldWithPath("totalElements").type(JsonFieldType.NUMBER).description("전체 게시글 갯수"),
+					fieldWithPath("last").type(JsonFieldType.BOOLEAN).description("마지막 페이지 여부"),
+					fieldWithPath("size").type(JsonFieldType.NUMBER).description("한 페이지에서 보여줄 게시글 갯수"),
+					fieldWithPath("numberOfElements").type(JsonFieldType.NUMBER).description("현재 페이지의 개시글 갯수"),
+					fieldWithPath("first").type(JsonFieldType.BOOLEAN).description("첫 페이지 여부"),
+					fieldWithPath("number").type(JsonFieldType.NUMBER).description("현재 페이지 번호"),
+					fieldWithPath("empty").type(JsonFieldType.BOOLEAN).description("리스트가 비어있는지 여부"),
+					fieldWithPath("sort.sorted").type(JsonFieldType.BOOLEAN).description("sort.sorted"),
+					fieldWithPath("sort.unsorted").type(JsonFieldType.BOOLEAN).description("sort.unsorted"),
+					fieldWithPath("sort.empty").type(JsonFieldType.BOOLEAN).description("sort.empty")
```
page를 쓰면 response가 굉장해지는군요 ㄷㄷ.. 대단하십니다..

neilsonT commented on src/main/java/com/devcourse/springbootboard/post/domain/Post.java:
```diff
@@ -0,0 +1,60 @@
+	@Column(name = "content", nullable = false)
+	private String content;
+
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(
+		name = "user_id",
+		referencedColumnName = "id",
+		foreignKey = @ForeignKey(value = ConstraintMode.NO_CONSTRAINT)
+	)
+	private User user;
```
User와 양방향이 아니라 단방향으로 할 때는 foreignKey를 지정해주는건가요??
아무 생각 없이 양방향으로 했었는데, 저도 단방향으로 바꿔봐야겠네요

samkimuel replied:
FK를 지정해주기보다는 FK에 default로 제약조건이 걸리더라구요! 저는 그 제약조건을 없애고 싶었어요.
게시글을 작성한 유저를 삭제하려고 할 때 물리 삭제(데이터베이스에서 삭제)를 하려고 했습니다. 그런데 제약조건 때문에 삭제되지 않더라구요.
그래서 41번째 라인을 추가하면서 게시글을 작성했던 유저도 삭제할 수 있게 만들었습니다.

추가로 이 부분에 대해서 데이터 무결성에 대해 공부하면 좋을 것 같아요! 저도 추상적으로만 알고 있어서 공부가 필요할 것 같네요🤣

neilsonT commented on src/main/java/com/devcourse/springbootboard/global/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,22 @@
+import com.devcourse.springbootboard.post.exception.PostNotFoundException;
+import com.devcourse.springbootboard.user.exception.UserNotFoundException;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
저는 extends 없는 ExceptionHandler를 사용했었는데,
extends ResponseEntityExceptionHandler를 하면 어떤게 다른지 궁금합니다!

samkimuel replied:
Spring MVC에서 기본으로 제공되는 Exception들의 처리를 간단하게 등록할 수 있다고 해서 상속을 받았는데, 실제로 사용하지는 않았네요,,

samkimuel replied:
ResponseEntityExceptionHandler을 확장해서 검증 Exception에 대한 handling을 하려고 했는데, 특정 예외에 대해 제가 계획한대로 구현할 수 없어서 해당 부분을 삭제했습니다. 
ResponseEntityExceptionHandler에 대해 충분히 알아보지 않고 사용했네요 ㅠㅠ

taehyunnn reviewed (commented):
명환님 코드 완전 가독성 좋은 것 같습니다~
테스트 코드 꼼꼼하게 작성하신거에 감탄하고 갑니다 ㅎㅎ
극락 코딩 화이팅이요~ 👍

samkimuel committed:
refactor: @RequestBody 검증, 검증 예외 처리, 알 수 없는 예외 처리

- @Valid 어노테이션을 통해 @RequestBody 객체 검증
- @Valid 시 발생하는 Exception handling
- BindingResult로부터 ResponseEntity에 담을 데이터만 추출하여 모델링

samkimuel committed:
chore: MYSQL 설정 일부를 환경변수로 변경

0923kdh commented on src/main/java/com/devcourse/springbootboard/post/dto/PostWriteRequest.java:
```diff
@@ -0,0 +1,29 @@
+@Getter
+public class PostWriteRequest {
+	@NotBlank(message = "제목을 입력해볼까요?")
+	@Size(min = 1, max = 20, message = "제목은 1~20자로 작성해볼까요?")
```
제목이 20글자인건 너무 짧지 않을까요 ?

0923kdh commented on src/main/java/com/devcourse/springbootboard/post/converter/PostConverter.java:
```diff
@@ -0,0 +1,33 @@
+			.build();
+	}
+
+	public Post convertPostWriteRequest(PostWriteRequest postWriteRequest, User user) {
```
이 메서드만 convert~ 인 이유가 있나요?

0923kdh commented on src/main/java/com/devcourse/springbootboard/post/service/PostService.java:
```diff
@@ -0,0 +1,80 @@
+import com.devcourse.springbootboard.user.repository.UserRepository;
+
+@Service
+public class PostService {
```
class level 에 @Transactional 을 안쓴이유가 있으신가요?

bnminji commented on src/main/java/com/devcourse/springbootboard/user/dto/UserSignUpRequest.java:
```diff
@@ -0,0 +1,23 @@
+import javax.validation.constraints.Positive;
+
+import lombok.Getter;
+
```
UserDto를 UserSignUpRequest, UserUpdateRequest/ UserResponse로 나누어 작성하셔서 확실히 가독성이 좋아진 것 같습니다! 근데 UserSignUpRequest 와 UserSignUpRequest의 내용이 동일하고, UserResponse 또한 User~Request에서 필드만 추가된 것이어서 코드가 중복되는 것 같다는 느낌이 들었습니다. 보통 관례적으로 DTO를 나누어서 작성하는 건지, 아니면 다른 방법(UserDto에서 constructor를 오버라이드 해서 사용, 상속 등..)이 존재하는 건지 궁금하네요..!

samkimuel commented on src/main/java/com/devcourse/springbootboard/global/error/ErrorResponse.java:
```diff
@@ -0,0 +1,64 @@
+	) {
+		this.status = status;
+		this.message = message;
+		this.errorClassName = errorClassName;
```
사실 저도 작성하면서 필요없다고 생각했었습니다ㅎㅎ,,, 일단 뽑을 수 있는 데이터를 다 뽑아서 보내보자! 하고 시도해본 것이어서 다희님 말씀대로 해당 부분을 로그로 관리하는 게 좋을 것 같습니다!

sangmin7648 reviewed (commented):
명환님 많이 배우고 갑니다!! 스처지나가듯 댓글 단거 확인해주세요

samkimuel commented on src/main/java/com/devcourse/springbootboard/user/dto/UserSignUpRequest.java:
```diff
@@ -0,0 +1,23 @@
+
+@Getter
+public class UserSignUpRequest {
+	@NotBlank(message = "이름을 입력해볼까요?")
```
어노테이션이 해당 필드의 검증을 하는 역할을 한다고 보시면 될 것 같습니다.
Request 에서는 데이터의 검증이 필요하다고 생각했어요. 그래서 어노테이션을 적용했습니다!
Response 에 데이터를 담을 때는 검증이 필요하지 않다고 생각했습니다. 그래서 적용하지 않았어요. 
민지님은 Response 검증이 필요하다고 생각하시나용??

samkimuel commented on src/main/java/com/devcourse/springbootboard/global/error/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,44 @@
+		return ErrorResponse.toResponseEntity(ErrorCode.INTERNAL_SERVER_ERROR, "", Collections.emptyList());
+	}
+
+	@ExceptionHandler(value = {PostNotFoundException.class})
+	protected ResponseEntity<ErrorResponse> handlePostNotFoundException(PostNotFoundException e) {
+		return ErrorResponse.toResponseEntity(e.getErrorCode(), "", Collections.emptyList());
+	}
+
+	@ExceptionHandler(value = {UserNotFoundException.class})
+	protected ResponseEntity<ErrorResponse> handleUserNotFoundException(UserNotFoundException e) {
+		return ErrorResponse.toResponseEntity(e.getErrorCode(), "", Collections.emptyList());
+	}
```
`PostNotFoundException`과 `UserNotFoundException`은 RuntimeException을 상속하는 커스텀 예외로 만들었고, java에서 제공하는 NotFoundException을 상속받지 않아서 해당 코드가 동작하지 않을 것 같습니다.

사실 저도 이 부분이 중복이라고 생각해서 PostNotFoundException과 UserNotFoundException의 상위 클래스를 만들어 해결해야할까 생각했는데, 괜찮은 방법일까요??

sangmin7648 replied:
404가 사실 가장 많이 일어나는 에러잖아요? 그래서 NotFoundException 클래스는 백엔드 API 만들면 거의 필수라고 생각합니다

samkimuel commented on src/main/java/com/devcourse/springbootboard/post/service/PostService.java:
```diff
@@ -0,0 +1,80 @@
+		return postConverter.toPostResponse(post);
+	}
+
+	@Transactional
```
여기도 굳이 붙이 필요는 없었지만 서비스 로직 내에서 명시해주고 싶어서 붙이게 되었습니다. 지우는 게 깔끔해 보일까요??

sangmin7648 replied:
깔끔함보다는 저는 필요가 없으면 없는게 낫다고 생각해서 여쭤봤습니다!

samkimuel replied:
상민님 의견에 동의합니다! 필요없는 부분을 굳이 넣을 필요는 없겠네요👍🏻

sangmin7648 commented on src/main/java/com/devcourse/springbootboard/global/error/ErrorResponse.java:
```diff
@@ -0,0 +1,64 @@
+
+@Getter
+public class ErrorResponse {
+	private final String status;
```
혹시 바디에 담고 싶으신 이유가 있나요??

sangmin7648 commented on src/main/java/com/devcourse/springbootboard/user/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@Getter
+public class UserNotFoundException extends RuntimeException {
```
이전에도 왜 NotFoundException을 만들어서 상속하지 않냐고 여쭤봤었는데. 그때 상속을 별로 안좋아한다고 말씀하셨던걸로 기억합니다!!

sangmin7648 replied:
네 저도 checked 예외를 사용하지 않다고 생각합니다. 언급하신 NotFoundException의 경우 javassit가 제공하는 클래스로 자바 기본 제공도 아니더라구요! 그래서 런타임 예외를 상속하는 NotFoundException을 만들어서 사용했습니다!

DevRunner21 commented on src/main/java/com/devcourse/springbootboard/post/converter/PostConverter.java:
```diff
@@ -0,0 +1,33 @@
+		return PostResponse.builder()
+			.postId(post.getId())
+			.title(post.getTitle())
+			.content(post.getContent())
+			.userId(post.getUser().getId())
```
여기서 항상 user조회 쿼리가 한번 더 나갈 것 같은데, 
페치조인으로 post를 조회할떄 user를 한번에 같이 가져오면 성능에 좋지 않을까요?
